### ISSUE TITLE
Hostport Manager is not IPv6 ready

### ISSUE THREAD

**Issue by saschagrunert** [kubernetes/kubernetes#80810]
**What happened**:

When creating a pod sandbox which contains host port mappings, then the hostport_manager.go will reject any non IPv4:
https://github.com/kubernetes/kubernetes/blob/8c8c4113f045a00bee611594e402dee07600812b/pkg/kubelet/dockershim/network/hostport/hostport_manager.go#L85-L87

**What you expected to happen**:

That the port gets added independently of the IP version.

**How to reproduce it (as minimally and precisely as possible)**:

1. Setup a CNI bridge plugin as follows:
```json
{
  "bridge": "cni0",
  "cniVersion": "0.3.0",
  "ipMasq": true,
  "ipam": {
    "routes": [
      { "dst": "0.0.0.0/0" },
      { "dst": "1100:200::1/24" }
    ],
    "ranges": [
      [{ "subnet": "172.0.0.0/16" }],
      [{ "subnet": "1100:200::/24" }]
    ],
    "type": "host-local"
  },
  "isGateway": true,
  "name": "crio",
  "type": "bridge"
}
```
2. Start CRI-O (current master is IPv6 ready)
3. Try to create the [hostport test sandbox](https://github.com/cri-o/cri-o/blob/master/test/testdata/sandbox_config_hostport.json):
```
> sudo crictl runp test/testdata/sandbox_config_hostport.json
FATA[0002] run pod sandbox failed: rpc error: code = Unknown desc = failed to add hostport mapping for sandbox k8s_podsandbox1_redhat.test.crio_redhat-test-crio_1(0fc2e87d7183c0ab0b92a5df7f443eeba92f3735c6c68d06e91d15b1c7ede333): invalid or missing IP of pod k8s_podsandbox
1_redhat.test.crio_redhat-test-crio_1_
```

**Environment**:
- Kubernetes version (use `kubectl version`): vendored 1.16.0-alpha.1, but it should also affect the master

**Comment by saschagrunert** [comment]
We should consider that in dual stack mode the rejection would make sense, but if we run a sandbox in IPv6 mode only then the hostport mapping could apply as well, right?

**Comment by aojea** [comment]
I think that it shouldn't reject any based on the ip family.

**Comment by saschagrunert** [comment]
> I think that it shouldn't reject any based on the ip family.

I agree, but in dual stack mode we would not be able to open the port twice for the same interface.

**Comment by aojea** [comment]
:thinking: don`t know if I understand your concern, but this will open a port in a different IP address, there is no collision if you open a port in 5.5.5.5:80 and [db0:23::2]:80, right?

**Comment by saschagrunert** [comment]
>  don`t know if I understand your concern, but this will open a port in a different IP address, there is no collision if you open a port in 5.5.5.5:80 and [db0:23::2]:80, right?

Yes, correct. Sorry for the confusion.

**Comment by BenTheElder** [comment]
> https://github.com/kubernetes/kubernetes/blob/8c8c4113f045a00bee611594e402dee07600812b/pkg/kubelet/dockershim/network/hostport/hostport_manager.go#L85-L87

dockershim is not used when using a CRI?

**Comment by BenTheElder** [comment]
cc @yujuhong @Random-Liu cam you comment on this?

**Comment by yujuhong** [comment]
Is this issue about dockershim or cri-o?

**Comment by aojea** [comment]
I´m confused too, the snippet of code refers to dockershim and indeed it seems to miss the IPv6 functionality. However, the test and the description refers to CRI-O that AFAIK doesn´t use dockershim, uses CRI.

@saschagrunert can you clarify?

**Comment by saschagrunert** [comment]
> I´m confused too, the snippet of code refers to dockershim and indeed it seems to miss the IPv6 functionality. However, the test and the description refers to CRI-O that AFAIK doesn´t use dockershim, uses CRI.
> 
> @saschagrunert can you clarify?

We use the hostport manager from the dockershim in CRI-O, so I think the issue will affect both (if the dockershim will support IPv6, too).

**Comment by aojea** [comment]
for completeness, I could talk with @saschagrunert and clarify that the problem is with dockershim.
They've found the issue in the CRI-O project, it seems they are importing the dockershim host port manager https://github.com/cri-o/cri-o/blob/71f12a46396dc57ebd135910ef8b65739ca7a2ce/server/sandbox_network.go#L77 and it fails because it doesn't support IPv6.
I think the issue is legit and my fix #80854 solves it, however, I let you decide since my knowledge in the runtimes area is very basic.
